src/dict.c

```diff
@@ -107,6 +114,65 @@ uint64_t dictGenCaseHashFunction(const unsigned char *buf, size_t len) {
     return siphash_nocase(buf,len,dict_hash_function_seed);
 }
 
+/* --------------------- dictEntry pointer bit tricks ----------------------  */
+
+/* The 3 least significant bits in a pointer to a dictEntry determines what the
+ * pointer actually points to. If the least bit is set, it's a key. Otherwise,
+ * the bit pattern of the least 3 significant bits mark the kind of entry. */
+
+#define ENTRY_PTR_MASK     7 /* 111 */
+#define ENTRY_PTR_NORMAL   0 /* 000 */
+#define ENTRY_PTR_NO_VALUE 2 /* 010 */
+
+/* Returns 1 if the entry pointer is a pointer to a key, rather than to an
+ * allocated entry. Returns 0 otherwise. This function can also be used to check
+ * if a key pointer can be stored directly in a dict hashtable bucket without a
+ * dictEntry. */
```

Comment:
we can document that if we use the no_value feature, keys must be odd, and a todo to fully implement the key_are_odd feature when that needs to change.

again, by fully implement i mean that for dicts with key_are_odd=0 we turn things around so that they're optimal as well.

Reply:
The TODO I'd like to add is to also add a `keys_are_even` to enable key-as-entry for dicts where all keys are even.

But what about dicts where keys can be a mix of odd and even? (String literals or other unaligned strings, or numeric keys.) To disable key-as-entry but still use dictEntry_no_value for these is a reasonable compromise IMHO.

Reply:
Done as described. Please check if it's acceptable.

Reply:
i'm not sure we'll ever need to support a case of a dict where keys are neither always odd or always even.
which is why i thought that this single `key_are_odd` decides between always odd and always even.
instead i now realize it decides between always odd and mixed, and we don't support the always even mode.

in my previous post i supported either fully implement all of this and support both (or all 3) modes, or just dropping all the complications and supporting just the one we use.

but anyway, i'm also ok with supporting the two modes you chose to support, as long it's properly documented what it does, and what's left as todo to reach the end goal.